### ISSUE TITLE
MacOS test failure with --build-boost

### ISSUE THREAD

**Issue by JasonCoombs** [libbitcoin/libbitcoin#1005]
on MacOS Version 10.13.5 (High Sierra) here is an easy-to-reproduce and easy-to-fix build test failure.

This failed (with libbitcoin/master branch and after many BOOST build warnings):

```
$ ./install.sh --build-boost --prefix=/Users/username/build-libbitcoin
...
  CXXLD    test/libbitcoin-test
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
FAIL: libbitcoin-test_runner.sh
========================================
   libbitcoin 4.0.0: ./test-suite.log
========================================

# TOTAL: 1
# PASS:  0
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: libbitcoin-test_runner.sh
===============================

libbitcoin-test(65380,0x7fff9ee78380) malloc: *** mach_vm_map(size=43071216221753344) failed (error code=3)
*** error: can't allocate region
*** set a breakpoint in malloc_error_break to debug
libc++abi.dylib: terminating with uncaught exception of type std::bad_alloc: std::bad_alloc
./libbitcoin-test_runner.sh: line 21: 65380 Abort trap: 6           ./test/libbitcoin-test ${BOOST_UNIT_TEST_OPTIONS} > test.log
FAIL libbitcoin-test_runner.sh (exit status: 134)

```

This succeeded:

(Note: instead of building BOOST 1.62 with --build-boost, I used my existing installation of 1.67)

```
$ grep BOOST_LIB_VERSION /usr/local/include/boost/version.hpp
//  BOOST_LIB_VERSION must be defined to be the same as BOOST_VERSION
#define BOOST_LIB_VERSION "1_67":
```

No errors, no warnings:

```
$ ./install.sh CFLAGS="-Og -g"
...
  CXXLD    test/libbitcoin-test
/Applications/Xcode.app/Contents/Developer/usr/bin/make  check-TESTS
PASS: libbitcoin-test_runner.sh
============================================================================
Testsuite summary for libbitcoin 4.0.0
============================================================================
# TOTAL: 1
# PASS:  1
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
Running 1936 test cases...
Platform: Mac OS
Compiler: Clang version 9.1.0 (clang-902.0.39.1)
STL     : libc++ version 5000
Boost   : 1.67.0

real	17m31.473s
user	22m36.492s
sys	6m2.003s
```

**Comment by evoskuil** [comment]
If the initial issue still occurs I wouldn't consider this resolved.

**Comment by thecodefactory** [comment]
Possibly related: gcc-5 builds the version3 stack cleanly.  Later versions of gcc such as gcc7/gcc8 (where gcc8 is current on my distro) generate a lot of boost warnings as well, but seems to be happier with boost 1.67.0 instead of the 1.62.0 provided by the install.sh

I don't follow the clang series versions as closely, but this may just be a similar issue.  @JasonCoombs You could try to replace the boost version in install.sh with 1.67.0 and see if the issue is resolved.  If so, would you also mind trying previous versions of clang to a point where it works cleanly with the original boost 1.62.0?

It's getting trickier to track the matrix here, particularly since erroring on more and more warnings (e.g. zmq 4.2.1 doesn't compile at all with our warning levels if you use gcc7/gcc8, so that has to be done out of band with gcc-5 or used elsewhere). I haven't begun to explore that matrix extended to OS X/clang, so information here is appreciated.

**Comment by JasonCoombs** [comment]
Update: building with --prefix works also, that doesn't appear to have been a factor in the test failure. However, I left out a detail, above, about needing to set

`DYLD_LIBRARY_PATH=/Users/username/build-libbitcoin/lib`

After the initial build attempt with --build-boost the library path was wrong which resulted in a different build error and no attempt to run the test.

**Comment by JasonCoombs** [comment]
I previously installed bx using brew

```
$ brew list bx
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/bin/bx
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/etc/bash_completion.d/bx
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/etc/libbitcoin/bx.cfg
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/include/bitcoin/ (124 files)
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/lib/libbitcoin-explorer.0.dylib
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/lib/pkgconfig/libbitcoin-explorer.pc
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/lib/ (2 other files)
/usr/local/Cellar/libbitcoin-explorer/3.5.0_3/share/doc/ (6 files)
```

And I have a running bs installation built from version 3

```
$ bs --version

Version Information:

libbitcoin-server:     3.4.0
libbitcoin-protocol:   3.4.0
libbitcoin-node:       3.4.0
libbitcoin-blockchain: 3.4.0
libbitcoin:            3.4.0
```

After making the mistake of NOT using --prefix to build version 4 from the master branch, my include files are now version 4 and I had to manually downgrade to the 3.4.0 libbitcoin binaries in /usr/local/lib in order for bx and bs 3.4.0 to continue working.

**Comment by evoskuil** [comment]
That’s no fun. I recommend always using a prefix dir. usr/local/ installs are the Linux version of DLL hell.

**Comment by JasonCoombs** [comment]
Perhaps, but it's a little more complicated than that because --prefix doesn't force the various build steps launched by install.sh to propagate the correct (custom) path, which is why I've left the version 4 include files in /usr/local/include/bitcoin for now despite the minor differences (the changes made to the include files for version 4 might be backwards-compatible when building version 3 unless something complains about the version number mismatch? it also might have caused some warnings about virtual functions being overloaded) it's clear there needs to be a clean way to have version 3 and version 4 co-exist, especially since we love our brew on MacOS so much that everyone is going to keep brewing the old batch of code even as some of us upgrade to libbitcoin version 4.

**Comment by JasonCoombs** [comment]
This is a separate issue, now get these warnings when building other libbitcoin v4 repositories:

```
In file included from test/p2p.cpp:25:
In file included from ./include/bitcoin/network.hpp:34:
./include/bitcoin/network/protocols/protocol_seed_31402.hpp:54:18: warning:
      'libbitcoin::network::protocol_seed_31402::start' hides overloaded virtual function [-Woverloaded-virtual]
    virtual void start(event_handler handler);
                 ^
./include/bitcoin/network/protocols/protocol_timer.hpp:58:18: note: hidden overloaded virtual function
      'libbitcoin::network::protocol_timer::start' declared here: different number of parameters (2 vs 1)
    virtual void start(const asio::duration& timeout, event_handler handler);
```

**Comment by evoskuil** [comment]
Depends on what you mean by “propagate”. I prefer to use static builds be because the proper dependencies get compiled into the binary. With dynamic linking the intent is to allow the linked to decide, with the goal of *sharing* dependencies, which is the root cause of problems.

**Comment by JasonCoombs** [comment]
Good point. Should the build instructions in the README urge building without dynamic libraries?

**Comment by evoskuil** [comment]
For separate issues I recommend creating delegate github issues ;).

**Comment by JasonCoombs** [comment]
On closer inspection, and possibly the cause of the failure with --build-boost the only build step invoked by install.sh that seems to have used /usr/local instead of my --prefix pertained to boost:

```
checking for boostlib >= 1.62.0... yes
configure: boost_CPPFLAGS : -I/usr/local/include
configure: boost_ISYS_CPPFLAGS : -isystem/usr/local/include
configure: boost_LDFLAGS : -L/usr/local/lib
configure: boost_BUILD_CPPFLAGS : -I/usr/local/include
```

Where I have 1.67 in /usr/local but --build-boost is getting and building 1.62

For each component and additional repository of libbitcoin the --prefix directory appears to correctly appear in the include and lib paths. However, /usr/local/include and /usr/local/lib are still present in those paths, so if there is ever a time when there is a file missing from the --prefix directory it might be found in /usr/local and kaplooey, DLL hell.

**Comment by evoskuil** [comment]
That’s the premise of library sharing, follow the search path and load based on the prioritization - not at compile time, but at load/link time. IOW the build cannot control it.

**Comment by JasonCoombs** [comment]
also, for each libbitcoin repository (e.g. libbitcoin-node) the README doesn't recommend using the install.sh script with --prefix but that's how I've successfully built them and now have v4 completely within my /Users/jasoncoombs/libbitcoin4 directory instead of conflicting with v3 in /usr/local

**Comment by evoskuil** [comment]
It's not that non-prefix (or dynamic) builds don't work, they just require the user to fully understand the consequences of building stuff into usr/local/, and similarly dynamic builds require a full understanding of linker conflicts.

**Comment by JasonCoombs** [comment]
By "user" you mean programmer, of course. And given enough time we do understand the consequences, but not knowing the architecture of libbitcoin as well as you do that knowledge isn't available to others until later. Should we revise the build instructions for each libbitcoin repository so that everyone builds using install.sh --prefix instead of the current (old) instructions? Windows users build with either VS or MinGW, presumably? There don't seem to be instructions for Windows now except on the libbitcoin/libbitcoin repository.

When following the Autotools (advanced users) instructions, how does the user include a --prefix parameter to autogen and config like with install.sh? This is a rhetorical question, we should include whatever the answer is in the suggested build steps even for advanced users to discourage /usr/local

**Comment by evoskuil** [comment]
Advanced users are on their own. The builds strictly follow autotools conventions, and advanced instructions provide the standard build steps to indicate that.

Full Visual Studio instructions are in the readmes of the two repos that produce executables, as well as libbitcoin. Intermediate libs have minimal instructions for all builds as otherwise they too easily get out of date and become inconsistent and eventually unreliable.

There is not yet mingw support, but that is in progress (i.e. code generation and functional CI) along wiht cmake. @pmienk

I can see recommending use of a prefix build and static linking, in order to avoid library conflicts (I have offered this advice for a long time). But I certainly wouldn't require or default it. The behavior is common to dynamically linking anything on linux (and osx/windows) with any dependency, not unique to libbitcoin architecture.

**Comment by JasonCoombs** [comment]
Agree, except that presently the lack of support by default for side-by-side installation of v3 and v4 libraries is an architecture choice we don't need to make. One that everyone is going to run into in practice when v4 is live and v3 is still in widespread use.

**Comment by JasonCoombs** [comment]
Is it correct that v4 is not backwards-compatible with v3 in terms of anything that does dynamic linking? If so, should we version the filenames?

**Comment by evoskuil** [comment]
This is why it's called Hell.

**Comment by JasonCoombs** [comment]
Hell on a Mac is at least enjoyable. :-)

**Comment by evoskuil** [comment]
The more closed (or smaller) the platform, the smaller the Hell. The same issues happen on macOS though. Linux people were naive to think that this was just a Windows problem years ago.

**Comment by evoskuil** [comment]
The answer is of course complete isolation. Hence the iOS app model.

**Comment by evoskuil** [comment]
Re: support for SxS install, all that is required is a prefix directory. We could build into a versioned subdirectory of the target directory by default. However this only works if each new build is versioned or if people stick to tagged releases only.

**Comment by evoskuil** [comment]
@JasonCoombs regarding this issue:
```
'libbitcoin::network::protocol_seed_31402::start' hides overloaded virtual function...
```
Would you mind opening a issue so we can track it?

**Comment by evoskuil** [comment]
@thecodefactory regarding this issue:

> zmq 4.2.1 doesn't compile at all with our warning levels if you use gcc7/gcc8, so that has to be done out of band with gcc-5 or used elsewhere...

Would you mind opening a issue so we can track it?

**Comment by thecodefactory** [comment]
> Would you mind opening a new issue so we can track it?

Sure.  I had sent info to @pmienk not so long ago when I found it.

**Comment by evoskuil** [comment]
Recommend closing based on discussion and related issue tracking.

**Comment by JasonCoombs** [comment]
I'm still looking at this in parallel with the --disable-shared issue #1007

https://travis-ci.org/JasonCoombs/libbitcoin/builds

**Comment by evoskuil** [comment]
As far as I can tell there is no --disable-shared issue.

**Comment by evoskuil** [comment]
Status?

**Comment by evoskuil** [comment]
Status?

**Comment by JasonCoombs** [comment]
see https://github.com/libbitcoin/libbitcoin/issues/1007#issuecomment-417428213

**Comment by evoskuil** [comment]
I'm closing this as won't fix because it's the intended behavior when using autotools, but if you think you have an improvement please issue a PR.
